### ISSUE TITLE
Confusing error "incomplete feed source value"

### ISSUE THREAD

**Issue by cueckoo** [cue-lang/cue#550]
_Originally opened by @shykes in https://github.com/cuelang/cue/issues/550_

### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version 0.3.0-alpha4 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?

- Latest alpha release: yes
- Latest stable release: no

### What did you do?

```
$ cat > test.cue <<'EOF'
import (
        pkg "foo.com/example"
)
settings: pkg.Settings
EOF
```

```
$ mkdir -p cue.mod/pkg/foo.com/example
```

```
$ cat > cue.mod/pkg/foo.com/example/example.cue <<'EOF'
package example

Settings: {
        {
                foo: true
        } | {
                foo: false
        }
}

// - Removing the `if` clause will fix the issue
// - Moving the `if` clause to top-level will fix the issue
// - Changing the `if` clause to reference a top-level field will fix the issue
Unrelated: {}
bla: {
        if Unrelated.anyKey != "" {}
}
EOF
```

```
$ cue-0.2.0 eval test.cue -e settings >/dev/null && echo "cue 0.2: OK"
$ cue-0.3.0-alpha-4 eval test.cue -e settings >/dev/null && echo "cue 0.3: OK"
```

### What did you expect to see?

```
cue 0.2: OK
cue 0.3: OK
```

### What did you see instead?
```
cue 0.2: OK
settings: incomplete feed source value example (type struct):
    ./test.cue:5:11
```

**Comment by cueckoo** [comment]
_Original reply by @myitcv in https://github.com/cuelang/cue/issues/550#issuecomment-704035681_

Smaller reproducer:

```
-- cue.mod/module.cue --
module: "mod.com"
-- cue.mod/pkg/foo.com/example/example.cue --
package example

Settings: {}
blah: Settings.anyKey
-- test.cue --
package x

import (
	pkg "foo.com/example"
)

settings: pkg.Settings
```

```
$ cue eval test.cue -e settings
settings: incomplete feed source value pkg (type struct):
    ./test.cue:7:11
```

`v0.2.0` appears to ignore/not consider the error in `blah: Settings.anyKey`, i.e. the reference to the non-existent field `anyKey`. `tip` is more strict on that, either by accident or design. Assuming that is by design, the actual error message itself gets lost along the way.

cc @mpvl

**Comment by cueckoo** [comment]
_Original reply by @shykes in https://github.com/cuelang/cue/issues/550#issuecomment-704042123_

Sorry my reproducer was misleading... Try it with a valid reference inside the `if` barrier, and it will still fail... I had changed it to an invalid key along the way, as an experiment, but that doesn’t seem to be the root cause.

Do you see the same behavior on your side?

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/550#issuecomment-705095652_

@shykes  The original example is indeed supposed to fail, although the error message could be improved.

Changing it to 
```
Unrelated: {anyKey: "str"}
bla: {
	if Unrelated.anyKey != "" {}
}
```
as you suggested, works fine for me. Do you have a concrete value of where you expect it to fail?

**Comment by cueckoo** [comment]
_Original reply by @shykes in https://github.com/cuelang/cue/issues/550#issuecomment-705099759_

Your example works fine for me too, although I could swear it failed before...

I did find a variation that fails for me. The default value seems to change the behavior.

```
package example

Settings: {
        {
                foo: true
        } | {
                foo: false
        }
	hello: string | *""
}

bla: {
        if Settings.hello != "" {}
}
```

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/550#issuecomment-705414195_

What do you mean with "the default value seems to change the behavior". Would be good to know what you did _exactly_ to reproduce the error you thought you had.

For me the behavior is still as expected. The key difference of this example with the previous one is that you are now looking up a field in `Settings`, which is _not_ a concrete struct. The embedded disjunction makes the result logically a disjunction as well that needs to be resolved before any field can be looked up in it. In general, CUE always fully evaluates a value from a lookup before any lookup is performed.

In other words, `Settings` is logically equal here to
```
Settings:  {
    foo: true
    hello: string | *""
} | {
    foo: false
    hello: string | *""
}
```

In this case, again, the reported error leaves much to be desired and should be improved.

Other than that, in all these cases it should indeed be an error. The v0.2.0 behavior was incorrect.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/550#issuecomment-735237156_

Analysis: the error in the package should indeed not carry over to the main package because it is an "incomplete" error. This means that the error could be resolved if the package were to be further instantiated. But as the main package is not using the reference, it is fine for it to not get this error.

Fix coming up.
